pkg/virt-launcher/virtwrap/nichotplug.go

```diff
@@ -184,6 +184,11 @@ func withNetworkIfacesResources(
 	domainSpec *api.DomainSpec,
 	f SetDomainFunc,
 ) (cli.VirDomain, error) {
+	onRootComplex := vmi.Annotations[v1.PlacePCIDevicesOnRootComplex]
+	if len(vmi.Spec.Domain.Devices.Interfaces) == 0 || onRootComplex == "true" {
+		return f(vmi, domainSpec)
```

Comment:
I'm contemplating about this one for a while now.

First, I want to admit that I seem to have came up with this terrible function. It does not look good at all, hiding the domain object creation.

The change here results in a logic change, where we will not always call the hook twice.
On one hand this makes sense, but at the same time I am a bit concern someone (sidecar logic authors) may assume we call it only once and depend on it.
In the past it was called many times, on each change to the libvirt metadata (which carried information we wanted to preserve), but now the only place we still call it more than once is here.

I feel we need to do something to compensate to balance this now. Perhaps adding text on the hook package and relevant methods to want about this. Even making sure it appears on the documentation/examples.

Lastly, I think we should add this "filter" outside and never call `withNetworkIfacesResources` if this is the case. The whole point of this function was to process the logic of calling things twice, but now it has a condition that it does not and makes it less focused.

WDYT?

Reply:
>  I am a bit concern someone (sidecar logic authors) may assume we call it only once and depend on it.

We will document https://github.com/kubevirt/community/pull/252 about calling the sidecar multiple times.

> Lastly, I think we should add this "filter" outside and never call withNetworkIfacesResources if this is the case. The whole point of this function was to process the logic of calling things twice, but now it has a condition that it does not and makes it less focused.
> 
> WDYT?

I like it. I'll change it.